src/services/goals.js

```diff
@@ -2389,3 +2390,166 @@ export async function destroyGoal(goalIds) {
     return 0;
   }
 }
+
+export async function createMultiRecipientGoalsFromAdmin(data) {
+  // const {
+  //   region, // string
+  //   group, // string, group ID
+  //   createReport, // bool
+  //   useCuratedGoal, // bool
+  //   creator, // string, user ID
+  //   templateId, // string, template ID
+  //   goalPrompts, // array, as follows
+  // [
+  //   {
+  //     promptId: 1,
+  //     title: 'FEI root cause',
+  //     fieldName: 'fei-root-cause'
+  //   }
+  // ],
+  // 'fei-root-cause': [ 'Community Partnerships', 'Family Circumstances' ],
+  // goalSource, // string
+  // goalDate, // string
+  // selectedGrants, // stringified JSON grant data
+  // } = data;
+
+  const grantIds = JSON.parse(data.selectedGrants).map((g) => g.id);
+
+  let templateId = null;
+  let isError = false;
+  let message = '';
+
+  if (data.useCuratedGoal && data.templateId) {
+    templateId = Number(data.templateId);
+  }
+
+  let template;
+
+  if (templateId) {
+    template = await GoalTemplate.findByPk(templateId);
+  }
+
+  let name = data.goalText;
+
+  if (template) {
+    name = template.templateName;
+  }
+
+  if (!name) {
+    isError = true;
+    message = 'Goal name is required';
+  }
+
+  let goalsForNameCheck = [];
+
+  if (!isError && grantIds.length > 0) {
+    goalsForNameCheck = await Goal.findAll({
+      attributes: ['id', 'grantId'],
+      where: {
+        grantId: grantIds,
+        name,
+      },
+    });
+  }
+
+  if (goalsForNameCheck.length) {
+    isError = true;
+    message = `Goal name already exists for grants ${goalsForNameCheck.map((g) => g.grantId).join(', ')}`;
+  }
+
+  if (isError) {
+    return {
+      isError,
+      message,
+    };
+  }
+
+  let endDate = null;
+
+  if (data.goalDate) {
+    endDate = data.goalDate;
+  }
+
+  const goals = await Promise.all(grantIds.map(async (grantId) => {
+    const goal = await Goal.create({
```

Comment:
Please use bulk create with the flag to return the id for each of the goals created. This will be significantly faster.